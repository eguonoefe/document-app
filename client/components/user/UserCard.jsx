import React from 'react';
import PropTypes from 'prop-types';
import parseDate from '../../utilities/parseDate';

/**
* User Card for viewing a single user
*
* @param {Object} props { firstName, lastName,
* email, Role, id, createdAt, onClick }
* @returns {String} HTML markup for the User Card
*/
const UserCard = (props) => {
  const { firstName, lastName, email, Role, id, createdAt, onClick } = props;
  return (
    <div>
      <a
        onClick={onClick}
        name={id}
        className="truncate"
        href="#!">{firstName} {lastName}</a>
      <p>{email}</p>
      {(Role.title === 'regularuser') && <p>Regular User</p>}
      {(Role.title === 'contributor') && <p>Contributor</p>}
      <p>Joined at: {parseDate(createdAt)}</p>
      <div className="divider" />
    </div>
  );
};

// Set Default Props
UserCard.defaultProps = {
  title: '',
  content: '',
  id: ''
};
// Set Props
UserCard.propTypes = {
  firstName: PropTypes.string,
  lastName: PropTypes.string,
  email: PropTypes.string,
  Role: PropTypes.object,
  createdAt: PropTypes.string,
  onClick: PropTypes.func,
  id: PropTypes.number
};

export default UserCard;
