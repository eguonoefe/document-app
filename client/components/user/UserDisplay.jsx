import React from 'react';
import { Modal } from 'react-materialize';
import PropTypes from 'prop-types';

import InputField from '../common/InputField';
/**
 * User Display component for creating a document
 *
 * @param {Object} props { onSubmit, onChange, firstName, lastName, email }
 * @returns {String} HTML markup for the User Display
 */
const UserDisplay = (
  { onSubmit, onChange, firstName, lastName, email }) => (
    <div className="document-form">
      <div className="center hover">
        <div className="card">
          <h5 className="center"> User Profile </h5>
          <div className="row profile">
            <div className="col l8 m8 s12">
              <InputField
                name="firstName"
                id="firstName"
                value={firstName}
                placeholder="First Name"
                className="validate form-design"
                type="text" onChange={onChange} />
            </div>
            <div className="col l8 m8 s12">
              <InputField
                name="lastName"
                id="lastName"
                value={lastName}
                placeholder="Last Name"
                className="validate form-design"
                type="text" onChange={onChange} />
            </div>
            <div className="col l8 m8 s12">
              <InputField
                name="email"
                id="email"
                value={email}
                placeholder="Email"
                className="validate form-design"
                type="email" onChange={onChange} />
            </div>
          </div>
          <div className="row user-action">
            <div className="col l6 m6 s12 input-field center">
              <Modal
                trigger={
                  <a
                    id="update-password"
                    className="waves-effect btn button-design"
                    data-target="passwordModal">
                    Change Password
                  </a>
                }
              >
                <div>
                  <h5 className="center"> Set Password</h5>
                  <div className="row">
                    <div className="col l12 m12 s12">
                      <InputField
                        name="oldPassword"
                        placeholder="Old Password"
                        className="validate form-design"
                        type="password" onChange={onChange} />
                      <InputField
                        name="password"
                        placeholder="Password"
                        className="validate form-design"
                        type="password" onChange={onChange} />
                    </div>
                    <div className="col l12 m12 s12">
                      <InputField
                        name="confirmPassword"
                        placeholder="Re-Type Password"
                        className="validate form-design"
                        type="password" onChange={onChange} />
                    </div>
                  </div>
                  <div className="center">
                    <button
                      id="password-button"
                      className="btn button-design modal-action modal-close"
                      type="submit" onClick={onSubmit}>
                      Save
                  </button>
                  </div>
                </div>
              </Modal>
            </div>
            <div className="col l6 m6 s12 input-field center">
              <button
                id="update-button"
                className="waves-effect btn button-design"
                type="submit" onClick={onSubmit} >
                Save
              </button>
            </div>
          </div>
        </div>
      </div >
    </div>
  );

UserDisplay.propTypes = {
  firstName: PropTypes.string,
  lastName: PropTypes.string,
  email: PropTypes.string,
  onChange: PropTypes.func.isRequired,
  onSubmit: PropTypes.func.isRequired
};

export default UserDisplay;
