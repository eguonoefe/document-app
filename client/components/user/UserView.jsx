import React from 'react';
import { Modal } from 'react-materialize';
import PropTypes from 'prop-types';
/**
 * User view component for viewing a User
 * Has update and delete modals for deleting and updating a User
 *
 * @param {Object} props { id, firstName, lastName, email, deleteUser }
 * @returns {String} HTML markup for the User view
 */
const UserView = ({ id, firstName, lastName, email, deleteUser }) => (
  <div className="card blue-grey darken-1">
    <div className="card-content white-text">
      <span className="card-title">{firstName} {lastName}</span>
      <div>{email}</div>
    </div>
    <div className="card-action">
      <div className="row">
        <div className="col l12 m12 s12">
          <Modal
            trigger={
              <a
                className="waves-effect btn button-design"
                data-target="passwordModal">
                Delete
                </a>}
            actions={
              <div>
                <button
                  className="btn btn-flat modal-action modal-close"
                  name={id}
                  onClick={deleteUser}>
                  Delete
                  </button>
                <button
                  className="btn btn-flat modal-action modal-close left">
                  Cancel
                  </button>
              </div>}>
            <div >
              <h5 className="center">
                Are you sure you want to delete the user</h5>
            </div>
          </Modal>
        </div>
      </div>
    </div>
  </div>
);

UserView.propTypes = {
  id: PropTypes.number.isRequired,
  firstName: PropTypes.string.isRequired,
  lastName: PropTypes.string.isRequired,
  email: PropTypes.string.isRequired,
  deleteUser: PropTypes.func.isRequired,
};

export default UserView;
