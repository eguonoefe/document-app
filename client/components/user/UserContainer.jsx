import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { withRouter } from 'react-router-dom';
import UserActions from '../../actions/UserActions';

import validate from '../../utilities/validate';
import UserDisplay from './UserDisplay';

const updateUser = UserActions.updateUser;
const viewUser = UserActions.viewUser;
/**
 * @class UserContainer
 * @extends {Component}
 */
export class UserContainer extends Component {
  /**
   * Creates an instance of UserContainer.
   * Binds the functions to the class
   * @param {Object} props -
   * @memberof UserContainer
   */
  constructor(props) {
    super(props);
    this.state = {
      firstName: '',
      lastName: '',
      email: '',
      roleId: ''
    };
    this.onSubmit = this.onSubmit.bind(this);
    this.onChange = this.onChange.bind(this);
  }

  /**
   * Makes an action call to get the current user information
   * Sets user information to the state
   * Initializes materialize modal
   *
   * @return {void}
   * @memberof UserContainer
   */
  componentDidMount() {
    this.props.viewUser(this.props.access.user.id)
      .then(() => {
        this.setState({
          firstName: this.props.user.user.firstName,
          lastName: this.props.user.user.lastName,
          email: this.props.user.user.email,
          roleId: this.props.user.user.roleId
        });
      });
    $('.modal').modal();
  }
  /**
  * Validates the UserDisplay input fields
  * Makes an action call to update user profile
  * Toasts the error/success message
  *
  * @return {void}
  * @memberof UserContainer
  */
  onSubmit() {
    try {
      const { valid } = validate.validateUpdateUser(this.state);
      if (!valid) {
        throw new Error('No field should be left blank');
      }
      this.props.updateUser(this.state, this.props.access.user.id)
        .then(() => {
          if (this.props.user.message) {
            return Materialize.toast(
              this.props.user.message, 2000,
              'indigo darken-4 white-text rounded');
          }
          Materialize.toast(
            'Profile updated!', 2000, 'indigo darken-4 white-text rounded');
          this.props.history.push('/dashboard');
        });
    } catch (err) {
      Materialize.toast(err.message, 3000,
        'indigo darken-4 white-text rounded');
    }
  }

  /**
  * Sets the event value to the state
  * @return {void}
  * @param {Object} event The event of the HTML component
  * @memberof UserContainer
  */
  onChange(event) {
    const field = event.target.name;
    this.setState({ [field]: event.target.value });
  }

  /**
  * Renders the User component
  * @returns {String} The HTML markup for the User
  * @memberof UserContainer
  */
  render() {
    return (
      <UserDisplay
        firstName={this.state.firstName}
        lastName={this.state.lastName}
        email={this.state.email}
        onChange={this.onChange}
        onSubmit={this.onSubmit} />
    );
  }
}

const mapPropsToState = state => (
  {
    access: state.access,
    user: state.user
  }
);

UserContainer.propTypes = {
  user: PropTypes.object.isRequired,
  access: PropTypes.object.isRequired,
  updateUser: PropTypes.func.isRequired,
  viewUser: PropTypes.func.isRequired,
  history: PropTypes.object.isRequired
};


export default connect(
  mapPropsToState, { updateUser, viewUser })(withRouter(UserContainer));
