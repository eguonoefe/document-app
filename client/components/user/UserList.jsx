import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Switch, Route, withRouter } from 'react-router-dom';
import PropTypes from 'prop-types';
import ReactPaginate from 'react-paginate';

import UserActions from '../../actions/UserActions';
import UserCard from './UserCard';
import UserView from './UserView';

const getAllUsers = UserActions.getUsers;
const viewUser = UserActions.viewUser;
const deleteUser = UserActions.deleteUser;
const searchUsers = UserActions.searchUsers;

/**
 * @class UserList
 * @extends {Component}
 */
export class UserList extends Component {
  /**
   * Creates an instance of UserList.
   * Binds class methods
   * @param {Object} props -
   * @memberof UserList
   */
  constructor(props) {
    super(props);
    this.state = {
      Users: [],
      User: {},
      offset: 0,
      pageCount: 0,
      searchTerm: ''
    };
    this.onClick = this.onClick.bind(this);
    this.deleteUser = this.deleteUser.bind(this);
    this.onSearch = this.onSearch.bind(this);
    this.handlePageClick = this.handlePageClick.bind(this);
  }

  /**
  * Makes action call to retrieve list of users
  *
  * @return {void}
  * @memberof UserList
  */
  componentDidMount() {
    this.updateUserList();
  }
  /**
   * Update the state if the props are changed
   *
   * @return {void}
   * @param {Object} nextProps -
   * @memberof UserList
   */
  componentWillReceiveProps(nextProps) {
    this.setState({
      Users: nextProps.UserList,
      pageCount: nextProps.pagination.pageCount,
      count: nextProps.pagination.count
    });
  }
  /**
  * Makes an action call to get all users
  * Sets list of users to the state
  *
  * @return {void}
  * @memberof UserList
  */
  updateUserList() {
    this.props.getAllUsers()
      .then(() => {
        this.setState({
          Users: this.props.UserList,
          count: this.props.pagination.count
        });
      });
  }
  /**
  * Makes an action call to view a user
  * Toasts the error/success message
  * Sets user to the state
  *
   * @param {Object} event -
   * @return {void}
   * @memberof UserList
   */
  onClick(event) {
    this.props.viewUser(event.target.name).then(() => {
      if (this.props.User.message) {
        return Materialize.toast(
          this.props.User.message, 2000,
          'indigo darken-4 white-text rounded');
      }
      this.setState({
        firstName: this.props.User.firstName,
        lastName: this.props.User.lastName,
        email: this.props.User.email,
        roleId: this.props.User.roleId
      });
      this.props.history.push(`${this.props.match.url}/view-user`);
    });
  }

  /**
   * Sets the searchterm to the state
   * Makes an action call to the search for a user
   * @param {Object} event -
   * @return {void}
   * @memberof UserList
   */
  onSearch(event) {
    this.setState({ searchTerm: event.target.value });
    this.props.searchUsers(event.target.value)
      .then(() => {
      }).catch(() => {

      });
  }

  /**
  * Makes an action call to delete a user
  * Toasts the error/success message
  *
  * @return {void}
  * @param {Object} event -
  * @memberof UserList
  */
  deleteUser(event) {
    this.props.deleteUser(event.target.name).then(() => {
      this.updateUserList();
      this.props.history.push(`${this.props.match.url}/all-users`);
    }).catch(() => {

    });
  }
  /**
  * Pagination for the list of users
  *
  * @return {void}
  * @param {Object} data -
  * @memberof UserList
  */
  handlePageClick(data) {
    const selected = data.selected;
    const limit = 5;
    const offset = Math.ceil(selected * limit);
    this.setState({ offset });
    this.props.getAllUsers(offset, limit).then(() => {
      this.setState({
        Users: this.props.UserList
      });
    });
  }

  /**
  * Renders the User List component
  *
  * @returns {String} HTML markup for the UserList
  * @memberof UserList
  */
  render() {
    const singleUser = this.state.Users.map(User => (
      <UserCard
        key={User.id} {...User}
        onClick={this.onClick} match={this.props.match} />
    ));
    return (
      <div className="document-list">
        <div className="container">
          <div className="row">
            <div className="col l12 m12 s12">
              <div className="col l5 m5 s12">
                <input
                  className="search"
                  type="text"
                  name="searchTerm"
                  placeholder="Search.."
                  onChange={this.onSearch} />
              </div>
            </div>
            <div className="document-list-view">
              <div className="col l5 m5 s12">
                <div className=" card-panel hoverable">
                  <h5>All Users</h5>
                  <div className="scrollable">
                    {singleUser}
                    {(this.state.count > 5) && <ReactPaginate
                      previousLabel={'previous'}
                      nextLabel={'next'}
                      pageCount={this.state.pageCount}
                      marginPagesDisplayed={1}
                      pageRangeDisplayed={3}
                      onPageChange={this.handlePageClick}
                      containerClassName={'pagination'}
                      subContainerClassName={'pages pagination'}
                      activeClassName={'active'}
                    />}
                  </div>
                </div>
              </div>
            </div>
            <div className="col l6 m6 s12">
              <Switch>
                <Route
                  // eslint-disable-next-line
                  path={`${this.props.match.url}/view-user`} render={() => {
                    if (!this.props.User.id) {
                      this.props.history.push(`${this.props.match.url}`);
                    }
                    return (
                      <UserView
                        id={this.props.User.id}
                        firstName={this.state.firstName}
                        lastName={this.state.lastName}
                        email={this.state.email}
                        roleId={this.state.roleId}
                        deleteUser={this.deleteUser} />);
                  }} />
              </Switch>
            </div>
          </div>
        </div>
      </div>
    );
  }
}

UserList.propTypes = {
  getAllUsers: PropTypes.func,
  viewUser: PropTypes.func,
  deleteUser: PropTypes.func,
  searchUsers: PropTypes.func,
  User: PropTypes.object,
  UserList: PropTypes.array,
  match: PropTypes.object,
  pagination: PropTypes.object,
  history: PropTypes.object
};

const mapPropsToState = state => (
  {
    UserList: state.user.userList,
    pagination: state.user.pagination,
    User: state.user.user
  }
);

export default connect(
  mapPropsToState, {
    getAllUsers, viewUser, deleteUser, searchUsers
  })(withRouter(UserList));
