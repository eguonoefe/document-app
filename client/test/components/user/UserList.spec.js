/* global expect jest test */
import React from 'react';
import { shallow, configure } from 'enzyme';
import toJson from 'enzyme-to-json';
import Adapter from 'enzyme-adapter-react-16';
import { UserList } from '../../../components/user/UserList';

configure({ adapter: new Adapter() });

describe('Document List', () => {
  const props = {
    UserList: [{ email: 'Manny' },
    { email: 'John' }, { email: 'move' }],
    user: {
      firstName: 'Eguono',
      id: 3,
      lastName: 'john',
      email: 'role',
      password: 'hello',
      confirmPassword: 'hello'
    },
    getAllUsers: jest.fn(() => {
      return Promise.resolve();
    }),
    viewUser: jest.fn(() => {
      return Promise.resolve();
    }),
    searchUsers: jest.fn(() => {
      return Promise.resolve();
    }),
    deleteUser: jest.fn(() => {
      return Promise.resolve();
    }),
    history: {
      push: jest.fn()
    },
    match: {
      url: ''
    },
  };
  const component = shallow(<UserList {...props} />);
  beforeEach(() => {
    global.Materialize = { toast: jest.fn() };
  });
  describe('Container', () => {
    test('should match snapshot of user list', () => {
      const tree = toJson(component);
      expect(tree).toMatchSnapshot();
    });
    test('onSearch function should run when called', () => {
      const spy = jest.spyOn(component.instance(), 'componentDidMount');
      component.instance().componentDidMount();
      expect(spy).toHaveBeenCalled();
      component.instance().onSearch(
        { target: { value: 'Manny', name: 'searchTerm' } });
      expect(component.state('searchTerm')).toEqual('Manny');
    });
    test('deleteDocument function should run when called', () => {
      const spy = jest.spyOn(component.instance(), 'deleteUser');
      component.instance().deleteUser({ target: { value: 3, name: 'name' } });
      expect(spy).toHaveBeenCalled();
    });
    test('componentWillRecieveProps will run if new props are added', () => {
      const spy = jest.spyOn(component.instance(), 'componentWillReceiveProps');
      const userList = [{ firstName: 'john' }];
      const pagination = { pageCount: '' };
      component.setProps({ userList, pagination });
      expect(spy).toHaveBeenCalled();
    });
    test('changeDocument function should run when called', () => {
      const spy = jest.spyOn(component.instance(), 'onClick');
      component.instance().onClick({
        target: { name: 1 }
      });
      expect(spy).toHaveBeenCalled();
    });
  });
});
