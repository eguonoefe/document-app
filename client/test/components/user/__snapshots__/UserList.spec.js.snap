// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Document List Container should match snapshot of user list 1`] = `
<div
  className="document-list"
>
  <div
    className="container"
  >
    <div
      className="row"
    >
      <div
        className="col l12 m12 s12"
      >
        <div
          className="col l5 m5 s12"
        >
          <input
            className="search"
            name="searchTerm"
            onChange={[Function]}
            placeholder="Search.."
            type="text"
          />
        </div>
      </div>
      <div
        className="document-list-view"
      >
        <div
          className="col l5 m5 s12"
        >
          <div
            className=" card-panel hoverable"
          >
            <h5>
              All Users
            </h5>
            <div
              className="scrollable"
            />
          </div>
        </div>
      </div>
      <div
        className="col l6 m6 s12"
      >
        <Switch>
          <Route
            path="/view-user"
            render={[Function]}
          />
        </Switch>
      </div>
    </div>
  </div>
</div>
`;
